ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java

```diff
@@ -288,14 +285,30 @@ private void clean(CompactionInfo ci, long minOpenTxnGLB, boolean metricsEnabled
       if (metricsEnabled) {
         Metrics.getOrCreateCounter(MetricsConstants.COMPACTION_CLEANER_FAILURE_COUNTER).inc();
       }
-      txnHandler.markFailed(ci);
-    } finally {
+      handleCleanerAttemptFailure(ci);
+    }  finally {
       if (metricsEnabled) {
         perfLogger.perfLogEnd(CLASS_NAME, cleanerMetric);
       }
     }
   }
 
+  private void handleCleanerAttemptFailure(CompactionInfo ci) throws MetaException {
+    long defaultRetention = getTimeVar(conf, HIVE_COMPACTOR_CLEANER_RETRY_RETENTION_TIME, TimeUnit.MILLISECONDS);
+    int cleanAttempts = 0;
+    if (ci.retryRetention > 0) {
+      cleanAttempts = (int)(Math.log(ci.retryRetention / defaultRetention) / Math.log(2)) + 1;
+    }
+    if (cleanAttempts >= getIntVar(conf, HIVE_COMPACTOR_CLEANER_MAX_RETRY_ATTEMPTS)) {
+      //Mark it as failed if the max attempt threshold is reached.
+      txnHandler.markFailed(ci);
+    } else {
+      //Calculate retry retention time and update record.
+      ci.retryRetention = (long)Math.pow(2, cleanAttempts) * defaultRetention;
```

Comment:
So assuming HIVE_COMPACTOR_CLEANER_MAX_RETRY_ATTEMPTS==5m,  we try at CQ_COMMIT_TIME + 5m then CQ_COMMIT_TIME + 5^2 minutes then CQ_COMMIT_TIME + 5^3 minutes?

Reply:
I'm okay with this, I just can't remember if the original intention was this or e.g. the 3rd attempt should be at CQ_COMMIT_TIME + 5 + 5^2 etc.

Reply:
Yes, the plan was to have exponential backoff, the wait time doubles in every iteration (`2^n * 5m` where `n` is the number of attempts). Originally I wanted to add the backoff value to the last retry attempt time, but that would need one more new field. So to keep it simple I add it to the submit time.